text/0080-datetime-extension.md

```diff
@@ -0,0 +1,484 @@
+
+The `datetime` type does not provide a way for a policy author to create a `datetime` from a numeric timestamp. One of the readable formats listed above must be used, instead.
+
+Values of type `datetime` have the following methods:
+
+- `.offset(duration)` returns a new `datetime`, offset by duration.
+- `.toDate()` returns a new `datetime`, truncating to the day, such that printing the `datetime` would have `00:00:00` as the time.
+- `.toTime()` returns a new `duration`, removing the days, such that only milliseconds since `.toDate()` are left. This is equivalent to `DT - DT.toDate()`
+
+Values of type `datetime` can also be used with the `-` operator.
+
+- `DT1 - DT2` returns the difference between `DT1` and `DT2` as a `duration`. (The inverse of `-` is `DT1.offset(duration)`)
+
+An invariant for the `-` operator is that when `DT1` is before `DT2` the resulting duration is negative.
+
+`DT1` may be negative (_i.e._ milliseconds _before_ `1970-01-01T00:00:00Z`) requiring that `-` be computed specially:
+
+```
+if DT1 < 0 && DT2 >= 0 then
+    result = -(DT1 - DT2)
+else
+    result = DT1 - DT2
+```
+
+There is not a use case for adding, or multiplying `datetime` values together. As such, `DT1 * DT2` and `DT1 + DT2` are errors.
+
+Values of type `datetime` can be used with comparison operators:
+
+- `DT1 < DT2` returns `true` when `DT1` is before `DT2`
+- `DT1 <= DT2` returns `true` when `DT1` is before or equal to `DT2`
+- `DT1 > DT2` returns `true` when `DT1` is after `DT2`
+- `DT1 >= DT2` returns `true` when `DT1` is after or equal to `DT2`
+- `DT1 == DT2` returns `true` when `DT1` is equal to `DT2`
+- `DT1 != DT2` returns `true` when `DT1` is not equal to `DT2`
+
+Equality is based on the underlying representation (see below) so, for example, `datetime("2024-08-21T") == datetime("2024-08-21T00:00:00.000Z")` is true. This behavior is consistent with the decimal extension function, where `decimal("1.0") == decimal("1.0000")` is also true.
+
+
+#### Representation
+
+The `datetime` type is internally represented as a `long` and contains a Unix Time in milliseconds. This is the number of non-leap seconds that have passed since `1970-01-01T00:00:00Z` in milliseconds. Unix Time days are always 86,400 seconds and handle leap seconds by absorbing them at the start of the day. Due to using Unix Time, and not providing a "current time" function, Cedar avoids the complexities of leap second handling, pushing them to the system and application.
+
+Negative Unix Time values represent the number of milliseconds before `1970-01-01T00:00:00Z`. 
+
+### Durations of Time (`duration`)
+
+The `duration(long, string)` function constructs a duration value. The string argument must be one of `"days", "hours", "minutes", "seconds", "milliseconds"`. Strict validation requires `long` and `string` to be literals, although evaluation/authorization support any appropriately-typed expressions. Values of type `duration` have the following methods:
```

Comment:
I also like the look of a single string that supports more complex durations. ISO does indeed provide a format for this, although it's a little ugly 😅 https://www.digi.com/resources/documentation/digidocs/90001488-13/reference/r_iso_8601_duration_format.htm 

Example: `P3Y6M4D` is 3 years, 6 months, and 4 days.

There is also a format for time intervals: https://en.wikipedia.org/wiki/ISO_8601#Time_intervals

Reply:
There are things about ISO 8601 intervals that complicate things:

1. .5Y is valid.
2. They are context sensitive, meaning `P1M` might mean 28, 29, 30, or 31 depending on how it's used.

I'd suggest we support: 

- `nD`: is `n * 86400000` milliseconds
- `nH`: is `n * 3600000` milliseconds
- `nM`: is `n * 60000` milliseconds
- `nS`: is `n * 1000` milliseconds
- `nm`: is `n` milliseconds

We can then eliminate the `P` and `T` leaders for the individual parts, as they don't provide much benefit to us.

- `5D3H15M10S500m` -> `443710500ms` if I did my math correctly.

We can say that it's inspired by ISO 8601 but unambiguous.

Reply:
This makes sense to me. I'm not a big fan of using (uppercase) M for minutes and (lowercase) m for milliseconds, but it's acceptable. Some alternatives are to 1) not support millisecond-level granularity, or 2) use multiple letters (e.g., `ms`) for milliseconds.

Reply:
And third option: use `.` for milliseconds, so your example above would become `5D3H15M10.500S`

Reply:
I would definitely prefer the readability of lowercase, and don't mind `ms` at all. So, perhaps then:

* `nd`: is `n * 86400000` milliseconds
* `nh`: is `n * 3600000` milliseconds
* `nm`: is `n * 60000` milliseconds
* `ns`: is `n * 1000` milliseconds
* `nms`: is `n` milliseconds

`n` should be signed, since negative durations are a thing we're supporting.

I guess I'm not opposed to `n` being fractional, though it seems that this format is expressive enough, and easily readable enough that it shouldn't need to be. (Though, we _do_ have Decimal support so precision isn't as problematic...)

I added a duration parser (and switched the tests to use it) in my [python implementation](https://gist.github.com/apg/bc793d42f38d73a7d84ffa0fc309419f), and I like the way it feels to use.

Reply:
I’d also vote against fractional expressions.  Let’s start simple.  We can always add them later if needed.

Reply:
Updated the RFC with all of this!